src/realm/sync/subscriptions.hpp

```diff
@@ -339,17 +339,42 @@ class SubscriptionStore : public std::enable_shared_from_this<SubscriptionStore>
     util::Optional<PendingSubscription> get_next_pending_version(int64_t last_query_version) const;
     std::vector<SubscriptionSet> get_pending_subscriptions() REQUIRES(!m_pending_notifications_mutex);
 
-    // Update the state of an existing subscription set. `error_str` must be set
-    // if the new state is Error, and must not be set otherwise. Many state
-    // transitions are not legal; see the state diagram on SubscriptionSet.
+    // Mark query_version as having received an error from the server. Will
+    // throw an exception if the version is not in a state where an error is
+    // expected (i.e. if it's already completed or superseded).
     //
-    // If set to State::Complete, this will erase all subscription sets with a
-    // version less than the given one.
+    // This should only be called internally within the sync client.
+    void set_error(int64_t query_version, std::string_view error_str);
+    // Mark query_version as having begun bootstrapping. This should be called
+    // inside the write transaction used to store the first set of changesets.
+    //
+    // This should only be called internally within the sync client.
+    void begin_bootstrap(const Transaction&, int64_t query_version);
+    // Mark query_version as having completed bootstrapping. This should be
+    // called inside the write transaction which removes the final pending changeset.
+    //
+    // This should only be called internally within the sync client.
+    void complete_bootstrap(const Transaction&, int64_t query_version);
+    // Roll query_version back to the Pending state if it is currently Bootstrapping.
+    // Has no effect if the bootstrap in progress is not the first boostrap for
+    // this subscription set.
     //
     // This should only be called internally within the sync client.
-    void update_state(int64_t version_id, SubscriptionSet::State state,
-                      std::optional<std::string_view> error_str = util::none)
-        REQUIRES(!m_pending_notifications_mutex);
+    void cancel_bootstrap(const Transaction&, int64_t query_version);
+    // Report that a download has completed, meaning that the active subscription
+    // set should advance to the Completed state if it is currently in the
+    // AwaitingMark state. Has no effect if it is in any other state.
+    //
+    // This should only be called internally within the sync client.
+    void download_complete();
```

Comment:
Add a line about server initiated bootstraps for these 4 functions - e.g.,
// Has not effect if this is a server initiated bootstrap for the active query version.